README.md
# BankAPI

## What is it?

BankAPI is a secure decentralized messaging system to send files/messages between banks and other types of financial institutions.

There is a reference implementation of the protocol which can be used off-the-shelf, which is production grade quality and is not only for testing and demonstration, although it fulfils those two roles as well.

The BankAPI protocol relies on OpenPGP (RFC 4880) + SHA512 + HTTPS. That's it, there is nothing more to it.

BankAPI is only a transmission protocol, and makes no assumptions of what kind of messages or file types banks will want to transfer.

## Key exchange

![layout](https://raw.githubusercontent.com/trustly/bankapi/master/doc/BankAPI%20Key%20Exchange.png)

## Protocol design

![layout](https://raw.githubusercontent.com/trustly/bankapi/master/doc/BankAPI%20Protocol%20Design.png)

The protocol has certain restrictions on the details and choices of algorithms
used for generating the OpenPGP messages.  The terminology used here conforms
to the one specified in [RFC2119](http://tools.ietf.org/html/rfc2119).  The
details specific to OpenPGP are documented in
[RFC4880](http://tools.ietf.org/html/rfc4880).

1. All OpenPGP messages **MUST** be symmetrically encrypted messages, where the
   session key is public-key encrypted.  The session key packet **MUST**
   contain the _Key ID_ of the key used to encrypt the message.  All messages
   **MUST** contain a signature packet.  Implementations **MAY** choose to
   generate one-pass signatures.

2. All parties **MUST** use a key length of 2048 bits for their RSA keys.
   Implementations **MAY** accept longer keys.  DSA and Elgamal **MUST NOT** be
   used.

3. The cipher algorithm **MUST** be one of AES128, AES192 and AES256.

4. If the OpenPGP message is a _compressed message_, the compression algorithm
   **MUST** be ZIP or ZLIB.

5. The OpenPGP message **MAY** be integrity protected.

6. The _literal data_ inside the OpenPGP message **MUST** be marked to contain
   binary data.  If a text or UTF-8 message is received, an implementation
   **MAY** choose to accept the message.  An implementation **MUST NOT**
   convert any _canonical line endings_ to their local counterpart.

7. The signature packet and (if present) the one-pass signature packet **MUST**
   contain the _Key ID_ of the key used to sign the message.  Implementations
   **SHOULD** only sign a message using a single key.  The hash algorithm used
   for signatures **MUST** be SHA-512.

8. The _ASCII Armor_, including _Armor Headers_, **MUST** be valid UTF-8.

## System design

![layout](https://raw.githubusercontent.com/trustly/bankapi/master/doc/BankAPI%20System%20Design.png)

The system is built on top of PostgreSQL and uses its pgcrypto contrib module.

Messages are encrypted and signed using RSA public-key cryptography.

Each bank generates a RSA key-pair consisting of a public and a secret key.

The public keys and API URLs are exchanged between the banks.

The task of exchanging public keys and API URLs is preferrably done over the
SWIFT network, by sending them in a normal MT999 SWIFT FIN-message.  This
allows banks to trust the validity of the public key and API URL, as the origin
of SWIFT messages can be trusted.

## Highlights

- RSA-encryption
- SHA512
- JSON-RPC / HTTPS
- Linux / PostgreSQL / Apache
- Open source / MIT-license
- Real-time request/response
- Instant delivery receipt
- Local archiving of files in PostgreSQL

## Overview of the interface

To send a message, the sending bank encrypts and signs a message using
Create\_Message(), and calls Get\_Message() to get the actual ciphertext
content of the message.  The ciphertext is then delivered to the receiving bank
by calling its Receive\_Message() API method, accessible at the API URL
provided by the receiving bank.  The API implementation must adhere to the
JSON-RPC standard and accept HTTPS POST.

The Receive\_Message() function returns a _delivery receipt_, which is a
cryptographic proof of the fact that the sender was able to verify the
signature contained within the message against the receiving bank's public key.
This allows the sender to be certain that the message was delivered and
decrypted successfully by the recipient.

The sending bank calls Decode\_Delivery\_Receipt() with the _delivery receipt_
as input to verify its validity.

## Database tables and columns

### Banks
- BankID : Unique identifier for the bank, preferrably the SWIFT BIC (PRIMARY KEY)
- Protocol : API procotol, example "https"
- Host : API host, Internet domain name, example "bank.com"
- Port : API port, Internet network port, example 443
- Path : API path, example "/api"
- Datestamp : Date/time when added to the table

### Files
- FileID : SHA512 hash of the Plaintext (PRIMARY KEY)
- Plaintext : The content of the file (UTF8)
- Datestamp : Date/time when added to the table

### Keys
- MainKeyID : Signature key (PRIMARY KEY)
- SubKeyID : Encryption key
- PublicKeyring : Contains public keys
- SecretKeyring : Contains secret keys (only set for your own bank, NULL for others)
- BankID : The BankID which the keys belong to
- Datestamp : Date/time when added to the table
- PrimaryKey : Only set to TRUE for one row per BankID, other keys are still valid, but this key is used when creating new messages

### Messages
- MessageID : SHA512 hash of the Cipherdata (PRIMARY KEY)
- FileID : The hash of the file send in the message
- FromBankID : The bank who sent the message
- ToBankID : The bank who received the message
- Cipherdata : The encrypted and signed message
- DeliveryReceipt : The encrypted and signed delivery receipt
- Datestamp : The date/time when the message was created
- Delivered : The date/time when the message was delivered

## Database functions

### Create\_Message(Plaintext, FromBankID, ToBankID)

Create\_Message() takes the plaintext context of a file as input and prepares
an encrypted _message_ signed by **FromBankID** which can be only decrypted by
**ToBankID**.  If the _message_ already exists, the existing data is reused.
An error is returned if the _secret key_ of **FromBankID** or the _public key_
of **ToBankID** can not be located in the database.

The return value is the _message id_ of the message.

### Get\_Message(MessageID)

Given a _message id_ as input, Get\_Message() returns the ASCII armored PGP
message.  This data can then be HTTPS POSTed to the Receive\_Message API method
of **ToBank**.

### Receive\_Message(Ciphertext text)

Receive\_Message() decrypts and verifies the signature of the input _message_.
The message's details are written to the _Messages_ table, and information
about the file and its plaintext contents are written into the _Files_ table.
This function will raise an error if:

1. The _secret key_ counterpart of the _public key_ used to encrypt the
message is not present in the table _Keys_.

2. The _public key_ counterpart of the _secret key_ used to sign the message
is not present in the table _Keys_.

3. The message is corrupt or the signature can not be verified.

The return value is a _delivery receipt_, which should be then used as the
contents of the response to the caller of the Receive\_Message API method.

### Decode\_Delivery\_Receipt(DeliveryReceipt text)

Decode\_Delivery\_Receipt() decrypts and verifies the given _delivery receipt_,
where the keys involved and the plaintext should match a row in Messages.  If
so, the Message has been successfully delivered, and the
Messages.DeliveryReceipt and Messages.Delivered columns are set.

## Installing and building .deb packages for installation

The repository includes scripts for building debian/ubuntu packages for the
bankapi and the required pgcrypto extensions (in PostgreSQL) in order to run
the code.

If you are running a different version of PostgreSQL then 9.3 then replace the
version in the commands below with the appropriate version. If you are running
the default version in Debian you can omit the version extension of the
packages names.

- sudo apt-get -y build-dep postgresql-9.3
- sudo apt-get -y install postgresql-server-dev-9.3
- make

This should produce two deb files, one for bankapi and one for the pgcrypto
extension.

### bankapi-0.1.deb

Package contains the command line tools for sending messages
(/usr/bin/bankapi), SQL needed for creating the bankapi in postgres and
installes a CGI as an endpoint for receving incoming communication messages.

The installation of the module installs and activates the CGI script. It will
be available under http://localhost/bankapi .

The installation does not finish the installation of the database but leaves
this as a manual task. Follow the following steps for a default database
installation:

- cd /usr/share/bankapi
- sudo -u postgres createdb bankapi
- sudo -u postgres psql --dbname=bankapi --single-transaction --no-psqlrc --file=install.sql
- Make sure www-data user can connect to the bankapi database, this can for
  instance be done by adding the following line where appropriate in the active
  pg\_hba.conf file: local www-data bankapi peer
- To optionally install the test bank data: sudo -u postgres psql --dbname=bankapi --single-transaction --no-psqlrc --file=testdata/index.sql

### postgresql-pgcrypto-openpgp-9.3.deb

This is an extension to the pgcrypto PostgreSQL extension. The extension only
installs with the extended functions, all functions normally in pgcrypto is
still accessed from the standard pgcrypto extension. The extension is called
pgcrypto\_openpgp.


debian/changelog
bankd (0.1) UNRELEASED; urgency=low

  * Initial release.

 -- per <per@dev1>  Tue, 05 Aug 2014 14:06:10 +0200


demo.sh
#!/bin/bash
export PGOPTIONS="-c client_min_messages=ERROR"

# Let's create two databases, one for TESTBANK001 and another for TESTBANK002.
# This will allow us to see exactly what is being written to each one of the databases.

dropdb TESTBANK001
dropdb TESTBANK002
createdb TESTBANK001
createdb TESTBANK002

psql -X -q -f install.sql -d TESTBANK001
psql -X -q -f install.sql -d TESTBANK002
psql -X -q -f testdata/test-banks.sql -d TESTBANK001
psql -X -q -f testdata/test-banks.sql -d TESTBANK002
psql -X -q -f testdata/test-publickeyrings.sql -d TESTBANK001
psql -X -q -f testdata/test-publickeyrings.sql -d TESTBANK002
psql -X -q -f testdata/test-secretkeyring-TESTBANK001.sql -d TESTBANK001
psql -X -q -f testdata/test-secretkeyring-TESTBANK002.sql -d TESTBANK002

# Use Create_Message() to encrypt/sign a new message.

MessageID=$(psql -P format=unaligned -t -c "
    SELECT MessageID FROM Create_Message(
        _MessageType := 'text/plain',
        _Plaintext   := 'This is a secret message from TESTBANK001 to TESTBANK002',
        _FromBankID  := 'TESTBANK001',
        _ToBankID    := 'TESTBANK002'
    );
" -d TESTBANK001)

# A unique MessageID is returned, which is a SHA512 hash of the encrypted/signed
# data stored in Messages.Cipherdata in which table the MessageID is the primary key.
# The Plaintext is written to Files.Plaintext, where the Files.FileID is a SHA512
# hash of the Plaintext and is also the primary key of the same table.

echo "MessageID: '$MessageID'"
echo

# Use Get_Message() to get the Messages.Cipherdata in ASCII armored format.
# This is the data we will transmit to the "to bank" in this case TESTBANK002.

Ciphertext=$(psql -P format=unaligned -t -c "
    SELECT Ciphertext FROM Get_Message(
        _MessageID := '$MessageID'
    );
" -d TESTBANK001)

echo "Ciphertext:"
echo "$Ciphertext"
echo

# The "to bank" receiving the message should make the Receive_Message()
# function accessible via HTTPS POST, allowing the "from bank" to call
# the function with the Ciphertext returned by Get_Message().

DeliveryReceipt=$(psql -P format=unaligned -t -c "
    SELECT DeliveryReceipt FROM Receive_Message(
        _Ciphertext := '$Ciphertext'
    );
" -d TESTBANK002)

echo "DeliveryReceipt:"
echo "$DeliveryReceipt"

# The Receive_Message() returns a DeliveryReceipt which is a proof that
# the "to bank" received the message, and consists of a SHA512 hash
# of the plaintext message, encrypted/signed by the "to bank",
# which allows the "from bank" to decrypt/verify the signature
# of the DeliveryReceipt, to be sure the message was delivered.

echo "Decode_Delivery_Receipt(DeliveryReceipt):"

psql -c "
    SELECT FileID, FromBankID, ToBankID FROM Decode_Delivery_Receipt(
        _DeliveryReceipt := '$DeliveryReceipt'
    );
" -d TESTBANK001

# Since the FileID is a SHA512 hash of the message, we know it will be:
# aac8c8a1bfc8e6ae97c8aaa132472f39cc2497db23bfae424b6ee52e4e34a92574e34f6d0cd713cdd630da72a4aa151030b3c1b3745ca393af119e45b929c943
# as the message in this case was 'This is a secret message from TESTBANK001 to TESTBANK002'.

# Now, let's look at the content of the tables Files and Messages in the two databases TESTBANK001 and TESTBANK002:

echo "TESTBANK001.Files:"
psql -P expanded -c "SELECT * FROM Files" -d TESTBANK001

echo "TESTBANK002.Files:"
psql -P expanded -c "SELECT * FROM Files" -d TESTBANK002

# As we can see, the message has been transferred successfully from TESTBANK001 to TESTBANK002.

# Now, let's look at Messages in both databases:

echo "TESTBANK001.Messages:"
psql -P expanded -c "SELECT * FROM Messages" -d TESTBANK001

echo "TESTBANK002.Messages:"
psql -P expanded -c "SELECT * FROM Messages" -d TESTBANK002



cgi/bankapi.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import sys
import json
import psycopg2
import psycopg2.extras
import re

pg_connect_str = "dbname=bankapi"
pg_conn = None

def respond(result=None, error=None, httperrorcode=None):
    global pg_conn

    if result is not None:
        print "Status: 200 OK"
        print "Content-Type: text/plain"
        print ""
        print result
    else:
        if httperrorcode is None:
            httperrorcode = 500
        if error is None:
            error = 'Unknown error'

        print "Status: %s %s" % (httperrorcode, error)

    if pg_conn is not None:
        pg_conn.close()

    sys.exit(0)

def error_log(s):
    sys.stderr.write("%s\n" % (s))

try:
    pg_conn = psycopg2.connect(pg_connect_str)
except psycopg2.Error as e:
    error_log("Failed to connect to database: error=%s, code=%s" % (e.pgerror, e.pgcode))
    respond(error='Failed to connect to database', httperrorcode=500)

try:
    indata = sys.stdin.read().strip()
except Exception as e:
    respond(error='Failed to decode input data', httperrorcode=400)

if indata is None or len(indata) == 0:
    respond(error='No data', httperrorcode=400)

cur = pg_conn.cursor(cursor_factory=psycopg2.extras.DictCursor)

sqldata = None
try:
    cur.execute('SELECT DeliveryReceipt FROM receive_message(%s)',
            [indata])

    sqldata = cur.fetchone()
    pg_conn.commit()
    respond(result=sqldata['deliveryreceipt'])
except psycopg2.InternalError as e:
    pg_conn.rollback()
    error_log("Call to Receive_Message failed: error=%s, code=%s" % (e.pgerror, e.pgcode))

    mg = re.match('ERROR:\s+(ERROR_\S+)', e.pgerror)
    if mg is not None:
        respond(error=mg.group(1), httperrorcode=500)
    else:
        respond(error='Receieve processing failed', httperrorcode=500)

except psycopg2.ProgrammingError as e:
    error_log("Call to Receive_Message failed with ProgrammingError: error=%s, code=%s" % (e.pgerror, e.pgcode))
    pg_conn.rollback()
    respond(error='Recieve processing failed', httperrorcode=500)

sys.exit(0)


demo-bankapi.sh
#!/bin/bash
set -e

# This script is based upon the databases configurd in demo.sh script. 
#
# Deliver a message from TESTBANK001 to TESTBANK002 using sendbankmessage. 
# In order for this to work it requires that there is a web server listening
# accepting incoming connections on http://localhost/bankapi that is connected
# to the database for TESTBANK002 here. It also requires that TESTBANK001
# database is initialized as in demo.sh
#
# Install the debian packages for bankapi to get the cgi listening, then modify
# /var/www/bankapi/bankapi.py to point to the TESTBANK002 database (dbname=TESTBANK002)
echo "Sending message using bankapi"
echo 'This is another secret message from TESTBANK001 to TESTBANK002' | \
    /usr/bin/bankapi send -d TESTBANK001 TESTBANK001 TESTBANK002 text/plain -


echo "Searching and adding the bankapi key for Trustly into TESTBANK001 database."
echo "Please respond to the questions interactivly"
echo "Import key with id: 2E4C6F05FC0156943ABA5EBA41F51DC2C04EA656"
/usr/bin/bankapi add -d TESTBANK001 TRLYSESSXXX

# List the messages present in both databases. The information in the two
# databases should really be the same as we update and keep all the message
# information in sync between the two. 
echo "List messages in the system of TESTBANK001"
/usr/bin/bankapi list -d TESTBANK001

echo "List messages in the system of TESTBANK002"
/usr/bin/bankapi list -d TESTBANK002

echo "The two lists of messages should be identical (apart from minor differences in timestamps)"

# Now fetch the last message in the list from TESTBANK002 database and display it using read
messagerow=$(/usr/bin/bankapi list -d TESTBANK002 | tail -n2 | head -n1)
echo "Displaying the last message in TESTBANK002: $messagerow"
messageid=$(echo "$messagerow" | cut -c21-30)
/usr/bin/bankapi read "$messageid"


cgi/api.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import sys
import json
import psycopg2
import psycopg2.extras
import re

pg_connect_str = "dbname=bankapi"
pg_conn = None

def respond(result=None, error=None, httperrorcode=None):
    global pg_conn

    if result is not None:
        print "Status: 200 OK"
        print "Content-Type: text/plain"
        print ""
        print result
    else:
        if httperrorcode is None:
            httperrorcode = 500
        if error is None:
            error = 'Unknown error'

        print "Status: %s %s" % (httperrorcode, error)

    if pg_conn is not None:
        pg_conn.close()

    sys.exit(0)

def error_log(s):
    print "s" >> sys.stderr

try:
    pg_conn = psycopg2.connect(pg_connect_str)
except psycopg2.Error as e:
    error_log("Failed to connect to database: error=%s, code=%, diag=%", e.pgerror, e.pgcode, e.diag)
    respond(error='Failed to connect to database', httperrorcode=500)

try:
    indata = sys.stdin.read()
except Exception as e:
    respond(error='Failed to decode input data', httperrorcode=400)

if indata is not None:
    cur = pg_conn.cursor(cursor_factory=psycopg2.extras.DictCursor)

    sqldata = None
    try:
        cur.execute('SELECT DeliveryReceipt FROM receive_message(%s)',
                [indata])

        sqldata = cur.fetchone()
        pg_conn.commit()
        respond(result=sqldata['deliveryreceipt'])
    except psycopg2.InternalError as e:
        pg_conn.rollback()
        error_log("Call to Receive_Message failed: error=%s, code=%, diag=%", e.pgerror, e.pgcode, e.diag)

        mg = re.match('ERROR:\s+(ERROR_\S+)', e.pgerror)
        if mg is not None:
            respond(error=mg.group(1), httperrorcode=500)
        else:
            respond(error='Receieve processing failed', httperrorcode=500)

    except psycopg2.ProgrammingError as e:
        error_log("Call to Receive_Message failed with ProgrammingError: error=%s, code=%, diag=%", e.pgerror, e.pgcode, e.diag)
        pg_conn.rollback()
        respond(error='Recieve processing failed', httperrorcode=500)
else:
    respond(error='Invalid method', httperrorcode=404)

sys.exit(0)
